src/sudoku.hs

```diff
@@ -0,0 +1,136 @@
+module Sudoku where
+import GHC.Arr (fill)
+import Data.List
+
+-- | A representation of the values in a sudoku from 0-9 TODO: SHOULD BE RENAMED!!!!!
+data SudVal = One | Two | Three | Four | Five | Six | Seven | Eight | Nine
+    deriving (Eq, Enum, Bounded)
+
+instance Show SudVal where
+    show One = "1"
+    show Two = "2"
+    show Three = "3"
+    show Four = "4"
+    show Five = "5"
+    show Six = "6"
+    show Seven = "7"
+    show Eight = "8"
+    show Nine = "9"
+
+data Line = Horizontal | Vertical
+    deriving (Eq)
+
+instance Show Line where
+    show Horizontal = "-"
+    show Vertical = "|"
+
+data Note = Candidate SudVal | Line Line SudVal
+    deriving (Eq)
```

Comment:
(After you explained to me the meaning of the `Line` constructor.)

A candidate line is not really a property of a single cell. It is described by: a line, a block, and a value. So this is not really the right place to remember this information.

Reply:
Regarding candidates: it would probably better to store the set of all possible candidates as a set (`Data.Set`).